llvm/include/llvm/ADT/STLExtras.h

```diff
@@ -996,13 +996,18 @@ class concat_iterator
 
   static constexpr bool ReturnsByValue =
       !(std::is_reference_v<decltype(*std::declval<IterTs>())> && ...);
-
+  static constexpr bool ReturnsConvertibleType =
+      (!std::is_same_v<
+           std::remove_reference_t<decltype(*std::declval<IterTs>())>,
+           ValueT> ||
```

Comment:
This is quite complicated and it's not clear to me what the negation operator applies to. Maybe hoist this to a helper struct and add tests with static_assert?